File: .skills/openclaw-skills/skills/gocybertrade/tview-indicators-analysis/scripts/prana_skill_client.js
#!/usr/bin/env node
import { randomUUID } from "node:crypto";
import { parseArgs } from "node:util";

const DEFAULT_BASE_URL = "https://claw-uat.ebonex.io";
const DEFAULT_SKILL_KEY = "100_indicators_analysis";
const AGENT_RUN_TIMEOUT_MS = 150000;
const AGENT_RESULT_RETRY_TIMES = 3;
const AGENT_RESULT_RETRY_INTERVAL_MS = 10000;

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function isTimeoutError(error) {
  if (!error) return false;
  return error.name === "TimeoutError" || error.name === "AbortError";
}

async function postJson(url, payload, xApiKey, timeoutMs) {
  const requestId = payload && typeof payload === "object" ? payload.request_id || "" : "";
  console.error(`[HTTP REQUEST] request_id=${requestId}`);
  const res = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": xApiKey,
    },
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(timeoutMs),
  });
  const text = await res.text();
  if (!res.ok) {
    const error = new Error(`HTTP ${res.status}: ${text}`);
    error.status = res.status;
    throw error;
  }
  return JSON.parse(text);
}

async function invokeAgentResultWithRetry(baseUrl, requestId, xApiKey) {
  const url = `${baseUrl.replace(/\/+$/, "")}/api/claw/agent-result`;
  let lastResult = { error: true, detail: "agent-result 尚未获取到结果" };

  for (let attempt = 1; attempt <= AGENT_RESULT_RETRY_TIMES; attempt += 1) {
    if (attempt > 1) {
      await sleep(AGENT_RESULT_RETRY_INTERVAL_MS);
    }
    try {
      const result = await postJson(url, { request_id: requestId }, xApiKey, AGENT_RUN_TIMEOUT_MS);
      lastResult = result;
      const status = String(result?.data?.status || "").toLowerCase();
      if (status !== "running") {
        return result;
      }
    } catch (error) {
      lastResult = { error: true, detail: `第 ${attempt} 次 agent-result 请求失败: ${error.message || error}` };
    }
  }

  return lastResult;
}

async function invokeAgentRun(baseUrl, skillKey, question, threadId, xApiKey) {
  const requestId = randomUUID();
  const url = `${baseUrl.replace(/\/+$/, "")}/api/claw/agent-run`;
  const payload = {
    skill_key: skillKey,
    question,
    thread_id: threadId,
    request_id: requestId,
  };

  try {
    return await postJson(url, payload, xApiKey, AGENT_RUN_TIMEOUT_MS);
  } catch (error) {
    if (isTimeoutError(error)) {
      return await invokeAgentResultWithRetry(baseUrl, requestId, xApiKey);
    }
    throw error;
  }
}

async function main() {
  let values;
  try {
    ({ values } = parseArgs({
      args: process.argv.slice(2),
      options: {
        question: { type: "string", short: "q" },
        "thread-id": { type: "string", short: "t", default: "" },
        help: { type: "boolean", short: "h", default: false },
      },
      allowPositionals: false,
    }));
  } catch (error) {
    console.error(error.message || error);
    process.exit(1);
  }

  if (values.help) {
    console.log('用法: node scripts/prana_skill_client.js -q "用户任务" [-t thread_id]');
    process.exit(0);
  }

  if (!values.question) {
    console.error("错误: 必须使用 -q / --question 提供用户需求任务描述。");
    process.exit(1);
  }

  const xApiKey = String(process.env.PRANA_SKILL_API_FLAG || "").trim();
  if (!xApiKey) {
    console.error("错误: 未检测到环境变量 PRANA_SKILL_API_FLAG。");
    process.exit(1);
  }

  try {
    const result = await invokeAgentRun(
      DEFAULT_BASE_URL,
      DEFAULT_SKILL_KEY,
      values.question,
      values["thread-id"] || "",
      xApiKey,
    );
    console.log(JSON.stringify(result, null, 2));
  } catch (error) {
    console.error(`请求失败: ${error.message || error}`);
    process.exit(2);
  }
}

await main();

File: .skills/openclaw-skills/skills/gocybertrade/prana-stock-scoring-analysis/scripts/prana_skill_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from __future__ import annotations

import argparse
import json
import os
import socket
import sys
import time
import uuid
import urllib.error
import urllib.request
from typing import Any, Dict

DEFAULT_BASE_URL = "https://claw-uat.ebonex.io"
DEFAULT_SKILL_KEY = "stock_scoring_analysis_v2"
AGENT_RUN_TIMEOUT_SEC = 150
AGENT_RESULT_RETRY_TIMES = 3
AGENT_RESULT_RETRY_INTERVAL_SEC = 10


def _headers(x_api_key: str) -> Dict[str, str]:
    return {
        "Content-Type": "application/json",
        "x-api-key": x_api_key,
    }


def _post_json(url: str, payload: Dict[str, Any], x_api_key: str, timeout: int) -> Dict[str, Any]:
    data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    request_id = payload.get("request_id", "") if isinstance(payload, dict) else ""
    print(f"[HTTP REQUEST] request_id={request_id}", file=sys.stderr)
    req = urllib.request.Request(
        url=url,
        data=data,
        method="POST",
        headers=_headers(x_api_key),
    )
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read().decode("utf-8"))


def _is_timeout_error(exc: BaseException) -> bool:
    if isinstance(exc, TimeoutError):
        return True
    if isinstance(exc, socket.timeout):
        return True
    if isinstance(exc, urllib.error.URLError):
        return isinstance(exc.reason, socket.timeout)
    return False


def invoke_agent_result_with_retry(base_url: str, request_id: str, x_api_key: str) -> Dict[str, Any]:
    url = base_url.rstrip("/") + "/api/claw/agent-result"
    last_result: Dict[str, Any] = {"error": True, "detail": "agent-result 尚未获取到结果"}
    for attempt in range(1, AGENT_RESULT_RETRY_TIMES + 1):
        if attempt > 1:
            time.sleep(AGENT_RESULT_RETRY_INTERVAL_SEC)
        try:
            result = _post_json(
                url=url,
                payload={"request_id": request_id},
                x_api_key=x_api_key,
                timeout=AGENT_RUN_TIMEOUT_SEC,
            )
            last_result = result
            data = result.get("data") if isinstance(result, dict) else None
            status = str((data or {}).get("status", "")).lower() if isinstance(data, dict) else ""
            if status != "running":
                return result
        except Exception as exc:  # noqa: BLE001
            last_result = {"error": True, "detail": f"第 {attempt} 次 agent-result 请求失败: {exc}"}
    return last_result


def invoke_agent_run(base_url: str, skill_key: str, question: str, thread_id: str, x_api_key: str) -> Dict[str, Any]:
    request_id = str(uuid.uuid4())
    run_url = base_url.rstrip("/") + "/api/claw/agent-run"
    run_payload = {
        "skill_key": skill_key,
        "question": question,
        "thread_id": thread_id,
        "request_id": request_id,
    }
    try:
        return _post_json(
            url=run_url,
            payload=run_payload,
            x_api_key=x_api_key,
            timeout=AGENT_RUN_TIMEOUT_SEC,
        )
    except Exception as exc:  # noqa: BLE001
        if _is_timeout_error(exc):
            return invoke_agent_result_with_retry(base_url, request_id, x_api_key)
        raise


def main() -> None:
    parser = argparse.ArgumentParser(
        description=(
            "调用 Prana 技能接口。参数参考 SKILL.md："
            "question=用户需求任务，thread_id=首次传空、后续传上一轮返回的 thread_id。"
        )
    )
    parser.add_argument("--question", "-q", required=True, help="用户需求任务描述，例如：帮我分析茅台股票的技术指标")
    parser.add_argument("--thread-id", "-t", default="", help="会话ID；首次调用传空，后续传上一轮返回的 thread_id")
    args = parser.parse_args()

    x_api_key = os.getenv("PRANA_SKILL_API_FLAG", "").strip()
    if not x_api_key:
        print("错误: 未检测到环境变量 PRANA_SKILL_API_FLAG。", file=sys.stderr)
        sys.exit(1)

    try:
        result = invoke_agent_run(
            base_url=DEFAULT_BASE_URL,
            skill_key=DEFAULT_SKILL_KEY,
            question=args.question,
            thread_id=args.thread_id,
            x_api_key=x_api_key,
        )
        print(json.dumps(result, ensure_ascii=False, indent=2))
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")
        print(f"HTTP {exc.code}: {body}", file=sys.stderr)
        sys.exit(2)
    except Exception as exc:  # noqa: BLE001
        print(f"请求失败: {exc}", file=sys.stderr)
        sys.exit(3)


if __name__ == "__main__":
    main()
